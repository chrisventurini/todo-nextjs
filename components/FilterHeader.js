import { connect } from 'react-redux';
import React, { Component } from 'react';

import { mapDispatchToFilterActions } from '../actions/filtering'

import Chip from '@material-ui/core/Chip';
import RadioButtonUnchecked from '@material-ui/icons/RadioButtonUnchecked';
import RadioButtonChecked from '@material-ui/icons/RadioButtonChecked';



class FilterHeader extends Component {

    state = {
        filterCompleted: true
    };

    constructor(props) {
        super(props);

        this.handleFilterClicked = this.handleFilterClicked.bind(this);
    }

    handleFilterClicked() {
        this.props.actions.toggleFilterCompleted();
        this.setState({
            filterCompleted: !this.state.filterCompleted
        });
    }

    render() {
        return (
            <div id="filter-header">
                <Chip
                    icon={this.state.filterCompleted ? <RadioButtonChecked /> : <RadioButtonUnchecked/> }
                    label="Filter Completed"
                    clickable
                    onClick={this.handleFilterClicked}
                    color="primary"
                />
            </div>
        )
    }

}

const mapState = state => state;

export default connect(mapState, mapDispatchToFilterActions)(FilterHeader);