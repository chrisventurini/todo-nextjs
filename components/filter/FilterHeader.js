import propTypes from 'prop-types';

import Chip from '@material-ui/core/Chip';
import RadioButtonUnchecked from '@material-ui/icons/RadioButtonUnchecked';
import RadioButtonChecked from '@material-ui/icons/RadioButtonChecked';
import withStyles from "@material-ui/core/styles/withStyles";

const completedFilterBtn = {
    marginTop: '7px',
    marginLeft: '25px'
};

const styles = {
    filterHeader: {
        backgroundColor: '#95B8D1',
        marginTop: '64px',
        height: '45px',
        width: '100%',
        boxShadow: '0px 2px 4px -1px rgba(0, 0, 0, 0.2), 0px 4px 5px 0px rgba(0, 0, 0, 0.14), 0px 1px 10px 0px rgba(0, 0, 0, 0.12)',
    },

    completedFilterBtn: {
        ...completedFilterBtn
    },

    completedFilterBtnDisabled: {
        ...completedFilterBtn,
        opacity: 0.6
    }
};

const FilterHeader = ({asyncCallsInProgress, classes, completedFiltered, onCompletedClick}) => {

    let icon = completedFiltered ? (<RadioButtonChecked />) : (<RadioButtonUnchecked/>),

        completedClasses = asyncCallsInProgress ? classes.completedFilterBtnDisabled : classes.completedFilterBtn;

    return (
        <div className={classes.filterHeader} >
            <Chip
                className={completedClasses}
                disabled={asyncCallsInProgress}
                icon={icon}
                label="Filter Completed"
                clickable={!asyncCallsInProgress}
                onClick={onCompletedClick}
                color="primary"
            />
        </div>
    )
};

FilterHeader.propTypes = {
    asyncCallsInProgress: propTypes.bool.isRequired,
    classes: propTypes.object.isRequired,
    completedFiltered: propTypes.bool.isRequired,
    onCompletedClick: propTypes.func.isRequired
};

export default withStyles(styles)(FilterHeader);

