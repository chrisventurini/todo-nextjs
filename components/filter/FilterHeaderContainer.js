import { connect } from 'react-redux';
import { Component } from 'react';
import propTypes from 'prop-types';

import FilterHeader from './FilterHeader';

import { mapDispatchToFilterActions } from '../../store/actions/filtering';


export class FilterHeaderContainer extends Component {

    static propTypes = {
        asyncCalls: propTypes.object.isRequired,
        filters: propTypes.object.isRequired
    };

    constructor(props) {
        super(props);

        this.state = {
            filterCompleted: props.filters.completed
        };

        this.handleCompletedClicked = this.handleCompletedClicked.bind(this);
    }

    handleCompletedClicked() {
        this.props.actions.toggleFilterCompleted();
        this.setState({
            filterCompleted: !this.state.filterCompleted
        });
    }

    render() {
        return (
                <FilterHeader
                    asyncCallsInProgress={this.props.asyncCalls.inProgress}
                    completedFiltered={this.state.filterCompleted}
                    onCompletedClick={this.handleCompletedClicked}
                />
        )
    }
}

export const _mapState = state => {
    return {
        asyncCalls: state.asyncCalls,
        filters: state.filters
    };
};

export default connect(_mapState, mapDispatchToFilterActions)(FilterHeaderContainer);

