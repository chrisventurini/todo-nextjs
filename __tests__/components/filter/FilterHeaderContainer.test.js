import mockStoreBuilder from "../../utils/mockStoreBuilder";
import {shallow} from "enzyme/build";
import FilterHeader from "../../../components/filter/FilterHeader";
import FilterHeaderContainer from "../../../components/filter/FilterHeaderContainer";
import React from "react";

describe('<FilterHeaderContainer />', () => {

    describe('rendering', () => {
        let SUTWrapper,
            storeData;

        beforeAll(() => {
            storeData = { asyncCalls: {}, filters: {} };
            let store = mockStoreBuilder(storeData);
            SUTWrapper = shallow(<FilterHeaderContainer store={store}/>).dive();
        });
        
        it('should have the stores asyncCalls as a property', () => {
            expect(SUTWrapper.instance().props).toHaveProperty('asyncCalls', storeData.asyncCalls);
        });

        it('should have the stores filters as a property', () => {
            expect(SUTWrapper.instance().props).toHaveProperty('filters', storeData.asyncCalls);
        });

        it('should have the stores todo actions as a property', () => {
            expect(SUTWrapper.instance().props).toHaveProperty('actions');
        });

        it('should render correctly', () => {
            expect(SUTWrapper).toMatchSnapshot();
        });

    });

    describe('When handling a completed filter clicked', () => {
        let SUTWrapper,
            toggleCompletedSpy,
            storeData;

        beforeAll(() => {
            storeData = { asyncCalls: {}, filters: { completed: true } };
            let store = mockStoreBuilder(storeData);
            SUTWrapper = shallow(<FilterHeaderContainer store={store}/>).dive();

            toggleCompletedSpy = jest.fn();
            SUTWrapper.instance().props.actions.toggleFilterCompleted = toggleCompletedSpy;

            SUTWrapper.find(FilterHeader).props().onCompletedClick();
        });

        it('should flip the filterCompleted state', () => {
            expect(SUTWrapper.state('filterCompleted')).toBe(false);
        });

        it('should dispatch', () => {
            expect(toggleCompletedSpy).toHaveBeenCalled();
        });

    });

});