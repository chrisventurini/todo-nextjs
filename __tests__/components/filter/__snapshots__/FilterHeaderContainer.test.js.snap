// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<FilterHeaderContainer /> constructing and rendering should render correctly 1`] = `
<FilterHeader
  asyncCallsInProgress={true}
  classes={
    Object {
      "completedFilterBtn": "FilterHeader-completedFilterBtn-2",
      "completedFilterBtnDisabled": "FilterHeader-completedFilterBtnDisabled-3",
      "filterHeader": "FilterHeader-filterHeader-1",
    }
  }
  completedFiltered={true}
  onCompletedClick={[Function]}
/>
`;
